### PR TITLE
MemoryAttributesTable: Remove Const Generic Parameter and Copy/Clone

### PR DESCRIPTION
**Pull request by os-d**
The MemoryAttributesTable definition in r-efi currently has a const generic parameter for the size of the `entry` array. However, this size can never be known at compile time: the MAT is generated many times during boot with varying sizes of entries, depending on how many RUNTIME_SERVICES_CODE and RUNTIME_SERVICES_DATA allocated sections in the memory map there are. Therefore, when writing code to produce the MAT, the only compile time value that can be used is 0 and the structure must be overallocated.

This commit updates the structure to remove the const generic and use 0 for the entry array size, as that is the only value that can be used. It also removes Clone and Copy from the derived traits, because it does not make sense to clone or copy a structure that is overallocated with a 0 length array: you would miss all the actual entries in the structure and if you just wanted to copy the header that is trivial to do.

Note that a dynamically sized type cannot be used here either: Rust can only store a pointer to that structure, at compile time you have to coerce the unsized element into a sized element, which has the same issue as using a const generic here: we don't know the size at compile time.

### CODE REVIEW

**Comment by os-d** [comment]
@dvdhrm can you review this PR?

**Comment by os-d** [comment]
@dvdhrm thanks for the review!

Keeping the const generic is fine, it is still usable as is, just a little extra boilerplate to always pass 0 (though you can rely on 0 being the default). I do think that the zero length array is the same concept and could be produced by autogenerated code, it is the same difference for Rust? But, if you feel strongly here, I can drop that part.

I do feel the Copy/Clone removal is more important and does lead to hard to catch bugs. When I was writing the code to use the memory attributes table, I ran into an implicit copy happening and so my 0 length array was filled with garbage values (yes, I should have caught this through auditing my code, but also it is easy to miss).

I understand your argument of wanting all structures to have the same semantics and not be aware of type intrinsics, but I do worry about the safety of that. I think any structure that can have a zero length array at the end (as many do) should not carry Copy/Clone because of the implicit pointing to garbage that can happen. I would assume that would be the advice from the Rust side, to not derive those traits for structures that are in essence headers for a greater range of memory. Deriving them here helps defeat some of the memory safety of Rust because now my memory is actually implicitly somewhere else and garbage. The other option (besides "be a better programmer") is to manually implement the traits to take the size component from the structure and copy the array as well. This would I assume be more implementation than you want in r-efi (and honestly probably not that useful).

**Comment by dvdhrm** [comment]
> Keeping the const generic is fine, it is still usable as is, just a little extra boilerplate to always pass 0 (though you can rely on 0 being the default). I do think that the zero length array is the same concept and could be produced by autogenerated code, it is the same difference for Rust? But, if you feel strongly here, I can drop that part.

I imagine `r-efi` to be effectively a `bindgen` generated set of modules for the UEFI spec (plus a very minor set of base helpers), so its result is meant to be predictable by users by just reading the spec. We do very limited editing ourselves, and try to pretend the output is machine generated.

If we changed `MemoryAttributesTable` to use a 0-sized array, I believe this would confuse more than it would help. Its structure would now differ from other trailing arrays, it would be non-obvious to deduce its definition from just reading the spec.

I fully agree that high-level helpers would likely not bother with static memory-tables. Yet I believe it is the right thing for `r-efi` to do.

> I do feel the Copy/Clone removal is more important and does lead to hard to catch bugs. When I was writing the code to use the memory attributes table, I ran into an implicit copy happening and so my 0 length array was filled with garbage values (yes, I should have caught this through auditing my code, but also it is easy to miss).
> 
> I understand your argument of wanting all structures to have the same semantics and not be aware of type intrinsics, but I do worry about the safety of that. I think any structure that can have a zero length array at the end (as many do) should not carry Copy/Clone because of the implicit pointing to garbage that can happen. I would assume that would be the advice from the Rust side, to not derive those traits for structures that are in essence headers for a greater range of memory. Deriving them here helps defeat some of the memory safety of Rust because now my memory is actually implicitly somewhere else and garbage. The other option (besides "be a better programmer") is to manually implement the traits to take the size component from the structure and copy the array as well. This would I assume be more implementation than you want in r-efi (and honestly probably not that useful).

We certainly did not put too much thought into adding `Copy` and `Clone` everywhere, so I am definitely interested in discussing possible adjustments. But please be aware that any such change is a breaking change, and we would want to apply it to all modules, not just a single type. We want all transpose guidelines to be the same for all modules, so if we discuss changing the auto-derive traits, we have to discuss them based on *all* modules, not just a single type (yet, the derives can still be based on structural properties of a type).

I understand that copying basic structures can easily lead to issues. This is not just related to trailing arrays, but also any kind of reference-counted members, or self-referential structures.

Generally, we don't want any abstractions or helpers in `r-efi` (other than the `base` module, which is required for type definitions). Hence, we consider all structures and types that we define as pure memory mappings. Meaning that any operations you perform on them with `r-efi` purely operate on the memory mapped data, rather than interpreting the data in any way. Cloning an object is equivalent to `memcpy()`, dropping an object is equivalent to discarding it. Everything else needs to be done by the high-level abstractions.

Now if we said we remove `Clone` from `MemoryAttributesTable` for safety reasons, it would give the impression `Clone` is safe on other structures where we derive it. And this is certainly not the case. It is safe from a rust-UB perspective, but not from a data perspective.

The reason we added `Copy` and `Clone` is to make data manipulation very easy (since that is what you do with `r-efi`). You can simply create copies of data, and then adjust it manually where needed. You can surely argue that some structures are likely never copied, but then again we want to have clear rules. For instance, creating a modified memory-descriptor can be done like this:

```rust
let old = efi::MemoryDescriptor { ... };

let new = efi::MemoryDescriptor {
    type = efi::LOADER_CODE,
    ..old
};
```

This copies a memory-descriptor and adjusts specific fields. This requires `Copy` for all types used by `MemoryDescriptor`, which might be other types defined by `r-efi`. We found this to be very convenient to implement abstractions for UEFI.

I do agree that we should have a better story for DSTs, but all ideas we had required the Rust allocator traits to be stable. Since we cannot know which allocator to use, we cannot easily provide helpers to allocate or copy DSTs. Maybe at some point we can create DSTs with thin-pointers in stable Rust, this would be ideal. But for now, this is not possible.

Long story short: What kind of transpose rule do you imagine? Drop `Copy`/`Clone` for all structures with trailing arrays? Or just drop `Copy`?

**Comment by os-d** [comment]
> Long story short: What kind of transpose rule do you imagine? Drop `Copy`/`Clone` for all structures with trailing arrays? Or just drop `Copy`?

Yes, I would propose dropping Copy/Clone from all structures with trailing arrays (specifically ZSTs/DSTs). I think the general rule for Rust should be that deriving Copy/Clone is safe and advisable for structures that are the same between a copy and the original. Structures with trailing DST arrays are not the same between the copy and the original (because the array is not part of the copy) and so we should not derive Copy/Clone for these structures.

I understand wanting to keep strict rules for r-efi and I think that makes sense. I believe that broadening this proposal to include all trailing array structures in r-efi keeps a strict rule (and makes safety sense). If you agree, I can update this PR to keep the const generic and update all trailing array structures in r-efi to remove Copy/Clone.

**Comment by dvdhrm** [comment]
I actually tried implementing dropping `Copy` everywhere, and it is really awkward with current stable Rust. We run into several problems, which eventually leaves a big mess of types that are Copy, and types which are not:

1) I really want `Clone`. It is really awkward to re-implement `Clone` if you have no access to amend the definition. But if we retain `Clone`, but drop `Copy`, then unions can no longer derive `Clone`. This can be fixed with a macro that derives `Clone` via `core::mem::copy_nonoverlapping()` and thus assumes `Copy` without actually having it part of the API. I implemented this, and it looks ok'ish, but lacks automatic verification ala `AssertParamIsCopy`. Quite non-invasive, though.

2) Union-fields must be either `Copy` or `ManuallyDrop<T>`, since their drop-handlers are not invoked. I think wrapping all union-fields in `ManuallyDrop` is quite invasive and annoying. So we'd have to make them all `Copy`, which might work out for now, but gets awkward if future unions include types we already decided not to make `Copy`.

I didn't try further. The fact that unions require `Copy` kinda defeats the purpose of dropping `Copy` everywhere.

I will be pushing an amendment for the transpose guidelines that documents why we derive Clone and Copy, but this is not meant to put an end to this discussion! I am just trying to document the status quo.

**Comment by dvdhrm** [comment]
(You can find my attempt here: https://github.com/dvdhrm/r-efi/tree/wip/nocp)

**Comment by os-d** [comment]
I see that 2 years ago I suggested also dropping `Clone` but today I am definitely aligned on keeping `Clone` :).

I did not realize the union restriction with `Copy` types, but that makes sense. I agree that is invasive, but honestly, I mind it less than accidental magic `Copy`s. That's one of my primary concerns with the Rust language, at least when I'm mostly working with folks who come from a C background and I worry they will get the wrong behavior when writing what seems correct to them.

I was playing around with it here: https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=99791af8313edc451796a5ce3b393063. I wasn't too aggrieved to write that and depending on which field you use, you may or may not need it. But, I fully understand if you don't want to take that breaking change/push that on consumers.

I did some LLM assisted searching and the UEFI spec has 48 unions currently with a total of 57 fields that are non-integral types (e.g. a structure instead of pointer/integer etc.). The majority of these are in the networking code. To me that blast radius doesn't feel huge, particularly since I'm assuming there has not been much Rust UEFI networking code written? And if so, I'd suggest not using the UEFI network stack :), but that's an entirely different discussion.

If you don't want to go the `ManuallyDrop` route, what about the route of having a policy that no structures with trailing [D|Z]STs shall be `Copy`? They retain `Clone`, since it is a manual operation. The union question shouldn't come into play there, the spec doesn't (I believe) and won't (I believe) have any unions that consistent of trailing zero size arrays. That would be very strange.

In any case, I had not realized this PR was still open, I will close this, happy to still have the discussion here, but I can also file a separate issue if you prefer.

**Comment by dvdhrm** [comment]
> I see that 2 years ago I suggested also dropping `Clone` but today I am definitely aligned on keeping `Clone` :).

That is great to hear!

> I did not realize the union restriction with `Copy` types, but that makes sense. I agree that is invasive, but honestly, I mind it less than accidental magic `Copy`s. That's one of my primary concerns with the Rust language, at least when I'm mostly working with folks who come from a C background and I worry they will get the wrong behavior when writing what seems correct to them.
> 
> I was playing around with it here: [...]

Maybe `ManuallyDrop` isn't that bad after all.

My biggest gripe is that it does not implement `From<T>`. But we could just implement `Into<ManuallyDrop<T>> for T` on our types. Then initialization can just be `MyUnion.foo = Foo { }.into()`, rather than requiring `MyUnion.foo = ManuallyDrop::new(Foo { })`. Since `Deref` is implemented, there would be little need for `ManuallyDrop` to be in-scope.

You feel like giving this a go? Given that this is semi-invasive, I am not sure how fast we could be merging this. However, I think this is a definite improvement over the status quo, and if Patina is on-board, I see little reason not to do this.

I mean, the change from `fn` to `unsafe fn` in v6.0.0 was invasive as well (and Patina didn't do the jump yet, right?), but welcome by users. So maybe this isn't too bad. After all, people are still trying to figure out the best way for UEFI ffi.

**Comment by os-d** [comment]
> > I see that 2 years ago I suggested also dropping `Clone` but today I am definitely aligned on keeping `Clone` :).
> 
> That is great to hear!
> 
> > I did not realize the union restriction with `Copy` types, but that makes sense. I agree that is invasive, but honestly, I mind it less than accidental magic `Copy`s. That's one of my primary concerns with the Rust language, at least when I'm mostly working with folks who come from a C background and I worry they will get the wrong behavior when writing what seems correct to them.
> > I was playing around with it here: [...]
> 
> Maybe `ManuallyDrop` isn't that bad after all.
> 
> My biggest gripe is that it does not implement `From<T>`. But we could just implement `Into<ManuallyDrop<T>> for T` on our types. Then initialization can just be `MyUnion.foo = Foo { }.into()`, rather than requiring `MyUnion.foo = ManuallyDrop::new(Foo { })`. Since `Deref` is implemented, there would be little need for `ManuallyDrop` to be in-scope.
> 
> You feel like giving this a go? Given that this is semi-invasive, I am not sure how fast we could be merging this. However, I think this is a definite improvement over the status quo, and if Patina is on-board, I see little reason not to do this.
> 

Yeah, I’ll take a crack at this, sounds reasonable to me.

> I mean, the change from `fn` to `unsafe fn` in v6.0.0 was invasive as well (and Patina didn't do the jump yet, right?), but welcome by users.

We just did the jump a week or two ago.

>So maybe this isn't too bad. After all, people are still trying to figure out the best way for UEFI ffi.

Absolutely. And questions like, we’ve always written specs with C in mind, what does it mean to write a spec with Rust in mind? :)

**Comment by dvdhrm** [comment]
>> [...] 
> Yeah, I’ll take a crack at this, sounds reasonable to me.

Very cool!

> > I mean, the change from `fn` to `unsafe fn` in v6.0.0 was invasive as well (and Patina didn't do the jump yet, right?), but welcome by users.
> 
> We just did the jump a week or two ago.

Again, very cool to hear!

> > So maybe this isn't too bad. After all, people are still trying to figure out the best way for UEFI ffi.
> 
> Absolutely. And questions like, we’ve always written specs with C in mind, what does it mean to write a spec with Rust in mind? :)

Oh, indeed!
